api/src/main/java/jakarta/enterprise/inject/spi/ProcessManagedBean.java

```diff
@@ -38,4 +41,16 @@ public interface ProcessManagedBean<X> extends ProcessBean<X> {
      * @throws IllegalStateException if called outside of the observer method invocation
      */
     public AnnotatedType<X> getAnnotatedBeanClass();
+
+    /**
+     * Returns a new {@link InvokerBuilder} for given method. The builder eventually produces an invoker
+     * for the given method.
+     * <p>
+     * The {@code method} must belongs to the bean being registered, otherwise an exception is thrown.
+     *
+     * @param method method of the bean being registered, must not be {@code null}
+     * @return the invoker builder, never {@code null}
+     * @since 4.1
+     */
+    public InvokerBuilder<Invoker<X, ?>> createInvoker(AnnotatedMethod<? super X> method);
```

Comment:
Should this be on `ProcessBean` instead of `ProcessManagedBean`?

I can't think of a reason you shouldn't be able to create an invoker for a method for a producer method, a producer field, a session bean or a synthetic bean.

Possibly you shouldn't be able to create one for an interceptor or decorator.

Reply:
> Should this be on `ProcessBean` instead of `ProcessManagedBean`?

I don't think so; the idea was to support methods declared on class-based beans.
@Ladicek I am pretty sure we discussed this but I don't recall what were the reasons for this exactly?

> I can't think of a reason you shouldn't be able to create an invoker for a method for a producer method, a producer field, a session bean or a synthetic bean.

A producer method has to be declared on a CDI bean - therefore you can create the invoker for it there.
A producer field isn't a method at all; a poor fit for method invokers :)
A session bean will still work, because `ProcessSessionBean` is a subtype of `ProcessManagedBean`
A synthetic bean - I am not sure I follow here. Anything can be a synthetic bean and the logic of these closely follows that of producers. Do you mean the return type of those beans? Note that it can be (just like with producers) even a primitive type.

> Possibly you shouldn't be able to create one for an interceptor or decorator.

Yes, that doesn't make much sense.

Reply:
> A producer method has to be declared on a CDI bean - therefore you can create the invoker for it there.
> A producer field isn't a method at all; a poor fit for method invokers :)

I'm not suggesting creating an invoker for the producer field or method itself, but for methods on the bean types of the bean created by the producer method or field. (i.e. usually the return type of the producer method, or the type of the producer field).

E.g. if I have a producer method
```java
@ApplicationScoped
public Connection getConnection() { /*...*/ }
```
then I should be able to create an invoker for `Connection.prepareStatement(String)`

Reply:
Possibly we need to clarify what's meant by "The {@code method} must belong to the bean being registered".

I assumed it meant you must be able to call the method on one of the bean types of the bean.

Reply:
So I originally wanted this to only apply to managed beans, because you know the exact class of the bean and the full set of its methods. It's also the only thing that I have actual use cases for.

It would probably be possible to extend this to producers, where you at least know the supertype of the bean and only the methods declared on that supertype (or inherited from further supertypes) could be invoked.

For synthetic beans, there's no way to obtain anything about the bean implementation class, so that's a no-go.

Reply:
> I'm not suggesting creating an invoker for the producer field or method itself, but for methods on the bean types of the bean created by the producer method or field. (i.e. usually the return type of the producer method, or the type of the producer field).

I see, thanks for clarification.
Currently, invokers are created (in Portable Extensions) from `AnnotatedMethod` - obtaining AM for arbitrary method of arbitrary type might be a little challenging whereas getting them from `ProcessManagedBean` is straightforward.

> Possibly we need to clarify what's meant by The `{@code method}` must belong to the bean being registered.

I think this is in place because otherwise you could observe `ProcessManagedBean<Foo>` and try to register an invoker for an `AnnotatedMethod` from bean `Bar`.

Reply:
I'll try to expand the word "belong", but @manovotn is right -- this is to prevent using `ProcessManagedBean<Foo>` to create an invoker for a method of a completely different class. How about:

```
The {@code method} must be declared on the bean class of the bean being registered
or inherited from a supertype of the bean class, otherwise an exception is thrown.
```

Reply:
Expanded the comment per above.

Reply:
That's much clearer :+1:

If the intention is that you can only create an Invoker for a managed bean, then the documentation on `BeanInfo` should probably be updated to reflect that.

Reply:
That's actually specified in `Invoker`, but I added a note to `BeanInfo` as well.